tlatools/src/tlc2/output/ExitStatus.java

```diff
@@ -0,0 +1,57 @@
+package tlc2.output;
+
+import static tlc2.output.EC.*;
+
+public class ExitStatus
+{
+
+    public static final int fromErrorCode(int errorCode)
+    {
+        final int exitStatus;
+        switch (errorCode) {
+            
+            case NO_ERROR:
+                exitStatus = 0;
+                break;
+
+            // System errors
+            case SYSTEM_CHECKPOINT_RECOVERY_CORRUPT:
+            case SYSTEM_DISK_IO_ERROR_FOR_FILE:
+            case SYSTEM_DISKGRAPH_ACCESS:
+            case SYSTEM_ERROR_CLEANING_POOL:
+            case SYSTEM_ERROR_READING_POOL:
+            case SYSTEM_ERROR_READING_STATES:
+            case SYSTEM_ERROR_WRITING_POOL:
+            case SYSTEM_ERROR_WRITING_STATES:
+            case SYSTEM_FILE_NULL:
+            case SYSTEM_INDEX_ERROR:
+            case SYSTEM_INTERRUPTED:
+            case SYSTEM_METADIR_CREATION_ERROR:
+            case SYSTEM_METADIR_EXISTS:
+            case SYSTEM_OUT_OF_MEMORY_TOO_MANY_INIT:
+            case SYSTEM_OUT_OF_MEMORY:
+            case SYSTEM_STACK_OVERFLOW:
+            case SYSTEM_STREAM_EMPTY:
+            case SYSTEM_UNABLE_NOT_RENAME_FILE:
+            case SYSTEM_UNABLE_TO_OPEN_FILE:
+                exitStatus = 2;
+                break;
+
+            // Violations
+            case TLC_ACTION_PROPERTY_VIOLATED_BEHAVIOR:
+            case TLC_INVARIANT_VIOLATED_BEHAVIOR:
+            case TLC_INVARIANT_VIOLATED_INITIAL:
+            case TLC_INVARIANT_VIOLATED_LEVEL:
+            case TLC_PROPERTY_VIOLATED_INITIAL:
+            case TLC_TEMPORAL_PROPERTY_VIOLATED:
+                exitStatus = 3;
+                break;
+        
+            default:
+                exitStatus = 255;
+        }
+
+        assert exitStatus >= 0 && exitStatus <= 255;
+        return exitStatus;
```

Comment:
@japgolly Why don't you just return the actual EC constant instead of mapping it to 0..255?

Reply:
0..255 is the range allowed by the shell. Higher bits are ignored and only the lowest 8 bits are used. This is dangerous because if you tried to return an error code of 1024 it would result in an exit status of 0 and falsely indicate success.

Reply:
Does this apply to all three platforms (Windows, Mac, and Linux)?

Reply:
For a platform-independent solution we could map all EC constants into the range 0..255.  I assume some consolidation of EC constants will reduce its number to allow us to introduce new EC constants in the future without exceeding 255.

Reply:
Not sure about Windows but I'd be surprised if Mac was different. Regardless, I don't think it'd be a good idea for users to have different behaviour on different platforms, consistency (and simplicity) would be better.

As to whether EC's need to be isomorphic to exit statuses, personally I don't see the need. Most apps just have a few exit status that represent a type of error (not the precise error itself). We could always add more mappings in future if needs arise; this function here makes it very easy.

Reply:
Works for Windows but not for macOS and Linux:

```java
public class RetVal {

	public static void main(String[] args) {
		final int retVal = Integer.valueOf(args[0]);
		System.out.println(retVal);
		System.exit(retVal);
	}
}
```

```
C:\Users\makuppe\class>java -cp . RetVal 2323976
2323976

C:\Users\makuppe\class>echo %ERRORLEVEL%
2323976
```

I'm not sold on the least common denominator.

Reply:
@quaeler Care to be the tiebreaker? ;-)

Reply:
I would pitch that the consumer of the TLC results should have the ability to know the exact reason for failure; we could restrict our exit codes to a handful (?representing thematic groupings of errors?) and then add extra output required to be parsed, but this seems like extra work on both sides for no apparent gain.
(So i vote for a mapping of all EC constants into [0,255])

Reply:
The mapping as-is is incomplete and thus inconsistent (e.g. violated assumptions missing). I propose to collect high-level groupings for the constants in EC and map those to 0..255:

1. Spec parsing errors
2. Config parsing errors
3. Semantic errors spec (evaluation failed)
3. Invalid combination/values of command-line parameters
4. Safety violations
5. Liveness violations
6. Assumption violations
5. Exceed supported state space size (number of states, length of traces, ...)
5. Exceeded system resources (disk full, ...)
6. System errors (cosmic ray, head crash)
7. Distributed TLC

Reply:
Cool. Thanks for confirming the Windows case.

Those are great suggestions for error groupings. I don't have the knowledge to write the mappings though, some of those EC constant names aren't very clear and I don't want to spend ages going back and forth in a PR; it's inefficient. The mappings as-is is complete in that it's a total function and all error codes map to a non-zero exit status. That's really the goal and scope of this issue. I suggest we merge this and move a more detailed mapping into a new issue to be completed by someone more experienced with this code base.

Reply:
I'm reluctant to commit a partial fix now knowing to release a breaking fix later.

Reply:
Sorry, I have to disagree that this is a partial fix. My goal (and 99% of use cases) are to have a non-zero exit status on failure for automating model checking. I can amend this PR so that it's always either 0 or 1 to make that clearer -- I created those samples groupings (2 & 3) more to demonstrate how to create groupings. Maybe that's why we having some misunderstanding. Then later someone who's willing to read through all error codes and decide how to group then for more detailed resolution can. It would only be a breaking change if it happened after a release, and if someone explicitly checked for an error status of 1 instead of non-zero (which would be unreasonable and silly).

What would you like me to do? How should we proceed?

Reply:
There are actually plans for a new Toolbox/TLC release in the foreseeable future Thus, I suggest to indeed reduce the scope of this PR to simply indicate 0 on success and 1 otherwise.  A future change can then add a more detailed (platform-dependent?) mapping.

Technically, lets remove ExitStatus.java and instead create a (total) function (e.g. ```errorConstantToExitValue``` in EC.java that maps NO_ERROR to exit value 0 and all other errors to 1 (with a meaningful TODO statement).

Lastly, it would be desirable if tlc2.tool.liveness.ModelCheckerTestCase could validate this functionality.

Reply:
No problem, sounds good. I'll make those changes within a few days.